### DISCUSSION TITLE
Why we can't use fetch("/some-api") in special files like page.tsx or route.ts

### DISCUSSION THREAD

**Discussion by nirvikpurkait** [vercel/next.js#60663] [Help] [answered]
### Summary

If we use `fetch("/api/some-api")` in special files like `page.tsx` or `route.ts`, we get an error, but if we have a custom component in the `page.tsx`, and inside the component there we fetch data using `fetch("/api/some-api")`, it works.

I searched about it on the internet and I got an answer that, those special files are not meant to fetch data. If that is true, then why `fetch("http://localhost:3000/api/some-api")` works on those special files even though they are also fetching data.

So basically why `page.tsx` or `route.ts` needs the full `baseUrl` to fetch data even from the same source.

**Comment by icyJoseph** [comment, marked as answer]
Some literature:

- https://nextjs-faq.com/fetch-api-in-rsc
- https://nextjs-faq.com/fetch-api-in-getserversideprops

The issue is really, how should fetch resolve the base url? In the browser, when you do fetch, `fetch('/some/resource')`, the browser adds the document origin as the base url to the request.

However, in a Node.js script, where there is no document, how should one resolve the base url? That is not a simple as one might think, and even if you used absolute paths to tell the server about its current URL, it still implies one additional HTTP round trip, which is very much unwanted.

A framework such as Nuxt.js, does support this use case though, but it is a concept of their server engine, called, Direct API calls:

- https://nuxt.com/docs/guide/concepts/server-engine#direct-api-calls

And it requires you to use their special `$fetch` function. Notice how their idea is to skip the one additional call I talk about above.

**Reply by nirvikpurkait** [reply]
Ok, I read the documentation and understood the reason, there it suggested to use the server-side logic directly.

Now what if I have two separate server-components (i.e `<FirstComp/>` and `<SecondComp/>`) with server-side logic directly, which need to be **revalidated** at different time interval, i.e-
`<FirstComp/>` needs to revalidated at every `60` seconds and,
`<SecondComp/>` needs to revalidate at every `10` seconds
and put them in a wrapper component `<WrapComp/>`. I could make use of route config to revalidate at lowest interval such as `export const revalidate = 10`

Wouldn't it make the `<FirstComp/>` to be ravalidated at every `10` seconds too?
